Support PasswordHash in User type

NikSays opened the pull request:
This closes #865, #866

**Note to reviewers:** remember to look at the commits in this PR and consider if they can be squashed
**Note to contributors:** remember to re-generate client set if there are any API changes

## Summary Of Changes

The Secret specified in User's ImportCredentialsSecret may contain the `passwordHash` field.
If the field is absent, the behavior is unchanged: the plain-text password is generated if necessary and stored in the credentials Secret.

If the `passwordHash` field is present, then `password` field is ignored and the resulting credentials Secret will contain only the hash. This is done to prevent ambiguous situations where the hash doesn't correspond to the password. It also prevents the plain-text password being accidentally used over the hash.

If the hash is an empty string, a passwordless user is created.

## Additional Context

Currently the hash must be generated with the SHA512 algorithm. A way to specify another hashing algorithm needs to be added. I will add it as another field in the secret, but I'm concerned the documentation for importCredentialsSecret will become too complicated.

#### Todo

- [ ] ~Specify hashing algorithm~
- [x] Add examples
- [x] Update documentation

Zerpet commented on controllers/user_controller.go:
```diff
@@ -40,6 +40,18 @@ const (
 
+// UserCredentials describes the credentials that can be provided in ImportCredentialsSecret for a User.
+// If the secret is not provided, a random username and password will be generated.
+type UserCredentials struct {
+	// Must be present if ImportCredentialsSecret is provided.
+	Username string
+	// If PasswordHash is an empty string, a passwordless user is created.
+	// If PasswordHash is nil, Password is used instead.
+	PasswordHash *string
+	// If Password is empty and PasswordHash is nil, a random password is generated.
+	Password string
+}
```
If this is not part of the API, I would prefer to move this struct to the `internal` package.

Zerpet reviewed (commented):
I had a quick look and it looks good overall. I want to review again the system test changes in more detail tomorrow.

Zerpet commented:
Thank you for contributing this feature. I like the high level logic proposed.

> [...]. A way to specify another hashing algorithm needs to be added. [...]

I'm not sure we need to, can you elaborate on this?

NikSays commented:
> I'm not sure we need to, can you elaborate on this?

I've realized that forcing SHA-512 is viable. It won't work only if someone needs to create the User with a SHA-256 hash without access to the original password, or if someone wants to use a custom hashing function provided by a plugin. Both scenarios seem unlikely. 
Implementing algorithm selection would require more effort, so until someone requests it or provides a use-case, SHA-512 should be enough.

NikSays commented:
I've added unit and system tests, updated the documentation, regenerated the asciidoc and the CRD. Also the branch was rebased on top of main, to simplify the commit history.

Zerpet commented:
The latest changes look good to me. I'll merge after CI passes 🚀

NikSays commented:
Hi, @Zerpet, sorry to bother you. Could you please restart the CI workflow, there was a timeout starting the cluster operator, seems like a one-off issue.

Zerpet commented:
Thank you!
